CGL-2360

gregorysprenger opened the pull request:
This PR contains the following:
- Confirm cGS pipeline launch inputs by batchdir.
- Set threshold for minimum file size after FASTQ demux in cGS pipeline.

<!--
# Clinical-Genomics-Laboratory/nf-cgl-cgs pull request

Many thanks for contributing to Clinical-Genomics-Laboratory/nf-cgl-cgs!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the dev branch, unless you're preparing a pipeline release.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/Clinical-Genomics-Laboratory/nf-cgl-cgs/tree/main/.github/CONTRIBUTING.md)
-->

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/Clinical-Genomics-Laboratory/nf-cgl-cgs/tree/main/.github/CONTRIBUTING.md)
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

gregorysprenger committed:
Added demux filesize checks; check if batch name is in run directory path

gregorysprenger committed:
Remove unnecessary script line continuation

gregorysprenger committed:
Check if 8 digit date in batch name matches Illumina run directory name

gregorysprenger committed:
Merge branch 'main' into CGL-2360

jasonwalker80 reviewed (commented):
<html><head></head><body>
<h1>Code Review for PR #7 (CGL-2360)</h1>
<h2>Summary</h2>
<p>This PR adds two validation checks to the Nextflow pipeline:</p>
<ol>
<li><strong>FastQ file size validation</strong> - ensures demultiplexed FastQ files are &gt;10MB</li>
<li><strong>Batch name date validation</strong> - verifies batch name date matches the Illumina run directory</li>
</ol>
<hr>
<h2>Detailed Review</h2>
<h3>1. <strong>FastQ Size Validation</strong> (<code>subworkflows/local/demultiplex.nf</code>)</h3>
<p><strong>Location</strong>: Lines 98-107</p>
<h4>✅ <strong>Strengths</strong></h4>
<ul>
<li>Good safety check to catch potential demultiplexing failures early</li>
<li>Conditional logic properly excludes validation samples</li>
<li>Clear error message indicating which file failed</li>
</ul>
<h4>⚠️ <strong>Issues &amp; Suggestions</strong></h4>
<p><strong>CRITICAL - Unused Channel:</strong></p>
<pre><code class="language-groovy">ch_verify_fastq_files = VERIFY_FASTQ_LIST.out.samples
    .map{ ... }
</code></pre>
<ul>
<li>This channel is assigned but <strong>never used or emitted</strong>, meaning the validation won't actually stop the pipeline</li>
<li>The <code>error()</code> will throw an exception, but the channel itself should be consumed</li>
</ul>
<p><strong>Recommendation:</strong></p>
<pre><code class="language-groovy">// Option 1: Add to workflow output or consume it
VERIFY_FASTQ_LIST.out.samples
    .map{
        meta, reads, fastq_list -&gt;
            reads.each{
                if (it.size() &lt;= (10 * 1024 * 1024)) {
                    error("FastQ file '${it}' is less than 10MB!")
                }
            }
            [meta, reads, fastq_list]  // Pass through the data
    }
    .set{ ch_verified_fastqs }

// Then use ch_verified_fastqs downstream instead of VERIFY_FASTQ_LIST.out.samples
</code></pre>
<p><strong>Style Issues:</strong></p>
<ul>
<li>The ternary operator <code>(it.size() &gt; (10 * 1024 * 1024)) ? true : error(...)</code> is awkward</li>
<li>Using a ternary just to return <code>true</code> or throw an error is unconventional</li>
</ul>
<p><strong>Better approach:</strong></p>
<pre><code class="language-groovy">reads.each{ fastq -&gt;
    if (fastq.size() &lt;= (10 * 1024 * 1024)) {
        error("FastQ file '${fastq.name}' is ${fastq.size()} bytes, less than 10MB minimum!")
    }
}
</code></pre>
<p><strong>Magic Number:</strong></p>
<pre><code class="language-groovy">10 * 1024 * 1024  // 10MB in bytes
</code></pre>
<ul>
<li>Consider defining this as a parameter or constant at the top of the file:</li>
</ul>
<pre><code class="language-groovy">def MIN_FASTQ_SIZE = 10 * 1024 * 1024  // 10MB
</code></pre>
<hr>
<h3>2. <strong>Batch Name Date Validation</strong> (<code>subworkflows/local/input_check.nf</code>)</h3>
<p><strong>Location</strong>: Lines 26-43</p>
<h4>✅ <strong>Strengths</strong></h4>
<ul>
<li>Excellent validation to prevent mismatched batch names and run directories</li>
<li>Clear, helpful error messages</li>
<li>Properly excludes validation samples</li>
<li>Good use of regex for date extraction</li>
<li>Helpful guidance in error messages about the <code>--validation_samples</code> flag</li>
</ul>
<h4>⚠️ <strong>Minor Suggestions</strong></h4>
<p><strong>Regex Specificity:</strong></p>
<pre><code class="language-groovy">def dateMatch = params.batch_name.find(/(\d{8})/)
</code></pre>
<ul>
<li>This captures any 8 digits, not necessarily a valid date</li>
<li>Consider a more specific regex:</li>
</ul>
<pre><code class="language-groovy">def dateMatch = params.batch_name.find(/(\d{4}(0[1-9]|1[0-2])(0[1-9]|[12]\d|3[01]))/)
</code></pre>
<ul>
<li>This ensures YYYYMMDD format with valid month (01-12) and day (01-31)</li>
</ul>
<p><strong>Variable Naming:</strong></p>
<pre><code class="language-groovy">def runDirName = new File(params.illumina_rundir).name
</code></pre>
<ul>
<li>Minor: Consider <code>runDirBasename</code> for clarity</li>
</ul>
<p><strong>Error Message Enhancement:</strong></p>
<pre><code class="language-groovy">error "Date in batch name ('${dateMatch}') does not match any part of the Illumina run directory name ('${runDirName}')..."
</code></pre>
<ul>
<li>Could add what was expected or show the parts that were checked:</li>
</ul>
<pre><code class="language-groovy">error "Date in batch name ('${dateMatch}') not found in Illumina run directory name ('${runDirName}'). Run directory parts checked: ${runDirName.split('_').join(', ')}. If this is a validation run, please set '--validation_samples' to true."
</code></pre>
<p><strong>Edge Case:</strong></p>
<ul>
<li>What if <code>params.illumina_rundir</code> is a relative path or doesn't exist? Consider adding:</li>
</ul>
<pre><code class="language-groovy">def runDir = new File(params.illumina_rundir)
if (!runDir.exists()) {
    error "Illumina run directory '${params.illumina_rundir}' does not exist!"
}
def runDirName = runDir.name
</code></pre>
<hr>
<h2>Overall Assessment</h2>

Aspect | Rating | Notes
-- | -- | --
Functionality | 🟡 Partial | FastQ validation won't execute properly; batch validation is solid
Code Quality | 🟡 Good | Clean but has the unused channel issue
Error Handling | 🟢 Excellent | Clear, actionable error messages
Testing Needed | 🔴 Critical | Needs testing to verify FastQ validation actually runs


<hr>
<h2>Required Changes</h2>
<ol>
<li><strong>Fix the FastQ validation</strong> - ensure the channel is actually consumed and affects the pipeline</li>
<li><strong>Test both validation paths</strong> with sample data</li>
<li>Consider the suggestions for improved date validation and error messages</li>
</ol>
<h2>Optional Improvements</h2>
<ol start="4">
<li>Extract magic numbers to named constants</li>
<li>Add more specific date regex validation</li>
<li>Add file existence check for illumina_rundir</li>
</ol>
<p>Would you like me to provide complete corrected code for either of these sections?</p></body></html>

gregorysprenger committed:
Better regex pattern match for dates

gregorysprenger committed:
Add min_fastq_size parameter and enforce size check for all input FastQ files unless validation samples parameter is true

gregorysprenger commented:
@jasonwalker80 I reworked the above so that file size checks are performed on all fastq files (fastq input and after demux). I also added a minimum file size parameter so that we can change the default minimum size if/when needed.

A small example data set was used to verify these changes. Please see the Box link on Jira.

jasonwalker80 reviewed (approved):
Looks good to me! See Claude comments below for potential, future improvements to consider. - JRW


<html><head></head><body><h1>Updated Review for PR #7 (CGL-2360)</h1>
<p>Great! I can see the PR has been <strong>significantly improved</strong> since our last review. The FastQ validation has been moved to a better location and the critical issue I identified has been fixed!</p>
<hr>
<h2>Summary of Changes</h2>
<ol>
<li><strong>FastQ file size validation</strong> - Moved to <code>input_check.nf</code> with configurable threshold</li>
<li><strong>Batch name date validation</strong> - Verifies batch name date matches Illumina run directory</li>
<li><strong>New parameter</strong> - <code>min_fastq_size</code> (default: 10MB) added to config</li>
</ol>
<hr>
<h2>Detailed Review</h2>
<h3>✅ <strong>Major Improvements Since Last Review</strong></h3>
<p><strong>1. FastQ Validation Location</strong> (<code>subworkflows/local/input_check.nf</code> lines 90-98)</p>
<ul>
<li><strong>Fixed!</strong> The validation is now inline during channel creation, so it will properly halt the pipeline</li>
<li>No more unused channel issue - the <code>error()</code> calls happen during the <code>.map()</code> operation</li>
<li>This is a much cleaner implementation</li>
</ul>
<p><strong>2. Configurable Threshold</strong></p>
<ul>
<li>Adding <code>min_fastq_size</code> parameter allows flexibility for different use cases</li>
<li>Well-documented in the schema with appropriate icon</li>
<li>Good default value (10MB)</li>
</ul>
<hr>
<h3>🟢 <strong>What's Working Well</strong></h3>
<p><strong>Batch Date Validation</strong> (lines 32-43)</p>
<pre><code class="language-groovy">def dateMatch = params.batch_name.find(/(\d{4}(0[1-9]|1[0-2])(0[1-9]|[12]\d|3[01]))/)
</code></pre>
<ul>
<li>Robust regex for YYYYMMDD format</li>
<li>Good error messages with helpful context</li>
<li>Proper escape hatch with <code>--validation_samples</code></li>
</ul>
<p><strong>FastQ Size Validation</strong> (lines 90-98)</p>
<pre><code class="language-groovy">def MIN_FASTQ_SIZE_BYTES = params.min_fastq_size * 1024 * 1024
if (R1.size() &lt; MIN_FASTQ_SIZE_BYTES) {
    error("FastQ file '${R1.name}' is ${R1.size()} bytes, less than ${params.min_fastq_size}MB minimum!")
}
</code></pre>
<ul>
<li>Clear error messages with actual file size</li>
<li>Checks both R1 and R2</li>
<li>Properly skipped for validation samples</li>
</ul>
<hr>
<h3>🟡 <strong>Minor Suggestions</strong></h3>
<p><strong>1. Error Message Formatting</strong>
The byte count could be more readable:</p>
<pre><code class="language-groovy">// Current:
error("FastQ file '${R1.name}' is ${R1.size()} bytes, less than ${params.min_fastq_size}MB minimum!")

// Suggested:
def actualSizeMB = R1.size() / (1024 * 1024)
error("FastQ file '${R1.name}' is ${actualSizeMB.round(2)}MB, less than ${params.min_fastq_size}MB minimum!")
</code></pre>
<p><strong>2. Consider Byte Size Display</strong>
Current error shows bytes (e.g., "10485760 bytes") which is harder to parse than "10MB"</p>
<p><strong>3. Validation Logic Consistency</strong>
Both validations check <code>!params.validation_samples</code> - good! But consider extracting this to a single check at the workflow level for maintainability.</p>
<hr>
<h3>📝 <strong>Documentation Additions Needed</strong></h3>
<p>Ensure these are documented:</p>
<ol>
<li><strong><code>min_fastq_size</code> parameter</strong> - what happens if set to 0? (should it disable validation?)</li>
<li><strong>Batch name format</strong> - document the expected YYYYMMDD format</li>
<li><strong>Error scenarios</strong> - add to troubleshooting guide</li>
</ol>
<hr>
<h3>✅ <strong>Testing Checklist</strong></h3>
<p>Before merging, verify:</p>
<ul>
<li>[ ] Pipeline fails with informative error when FastQ &lt; 10MB</li>
<li>[ ] Pipeline passes when FastQ &gt; 10MB</li>
<li>[ ] Validation samples bypass both checks with <code>--validation_samples</code></li>
<li>[ ] Custom <code>--min_fastq_size</code> values work correctly</li>
<li>[ ] Batch name date mismatch produces clear error</li>
<li>[ ] Missing date in batch name produces clear error</li>
<li>[ ] Edge case: FastQ exactly at threshold (10MB)</li>
</ul>
<hr>
<h2>Overall Assessment</h2>

Aspect | Rating | Notes
-- | -- | --
Functionality | 🟢 Excellent | Both validations will properly halt pipeline
Code Quality | 🟢 Excellent | Clean, maintainable implementation
Error Handling | 🟢 Excellent | Clear, actionable error messages
Configurability | 🟢 Excellent | Parameterized threshold is great


<hr>
<h2>Verdict: <strong>APPROVED with minor suggestions</strong> ✅</h2>
<p>This is a significant improvement over the initial implementation! The critical issue from my previous review has been resolved. The only suggestions are cosmetic improvements to error messages.</p>
<p><strong>Required before merge:</strong> None (code is functional)<br>
<strong>Nice to have:</strong> Improve error message readability by showing MB instead of bytes</p>
<p>Would you like me to provide the specific code snippets for the suggested error message improvements?</p></body></html>
